text/0000-remove-unnecessary-allocator-usage.md

```diff
@@ -0,0 +1,202 @@
+# RFC-0000: Remove unnecessary host functions allocator usage
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 2023-07-04                                                                                  |
+| **Description** | Add alternatives to host functions that make use of the allocator when unnecessary          |
+| **Authors**     | Pierre Krieger                                                                              |
+
+## Summary
+
+Add new versions of host functions in order to avoid using the allocator when the buffer being allocated is of a size known at compilation time.
+
+## Motivation
+
+The heap allocation of the runtime is currently controlled by the host using a memory allocator on the host side.
+
+The API of many host functions consists in allocating a buffer. For example, when calling `ext_hashing_twox_256_version_1`, the host allocates a 32 bytes buffer using the host allocator, and returns a pointer to this buffer to the runtime. The runtime later has to call `ext_allocator_free_version_1` on this pointer in order to free the buffer.
+
+Even though no benchmark has been done, it is pretty obvious that this design is very inefficient. To continue with the example of `ext_hashing_twox_256_version_1`, it would be more efficient to instead write the output hash to a buffer that was allocated by the runtime on its stack and passed by pointer to the function. Allocating a buffer on the stack in the worst case scenario simply consists in decreasing a number, and in the best case scenario is free. Doing so would save many Wasm memory reads and writes by the allocator, and would save a function call to `ext_allocator_free_version_1`.
+
+After this RFC, only the following functions have no exact equivalent that doesn't use the allocator:
+
+- `ext_storage_get`
+- `ext_default_child_storage_get`
+- `ext_storage_next_key`
+- `ext_default_child_storage_next_key`
+- `ext_crypto_ed25519_public_keys`
+- `ext_crypto_sr25519_public_keys`
+- `ext_crypto_ecdsa_public_keys`
+- `ext_offchain_network_state`
+- `ext_offchain_local_storage_get`
+- `ext_offchain_http_response_wait`
+- `ext_offchain_http_response_headers`
+- `ext_offchain_http_response_read_body`
```

Comment:
Note that this is to the extent of my knowledge. The list of host functions that are used in production (as opposed to the ones that are just prototypes) is not super well tracked.